ENH: adds constraints for differential evolution

andyfaff opened the pull request:
Adds constraints for `differential_evolution` following [Lampinen](https://pdfs.semanticscholar.org/088e/60df2694230e8e9e841e19ec218cddba54fe.pdf)
Depends on #9990 being merged first.
The first commit adds the machinery, tests are being created.

andyfaff commented:
TODO:

  - [ ] what happens if there is more than 1 constraint presented and each of the constraints has a different M?

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
         self.bounds = (lb, ub)
         self.keep_feasible = keep_feasible
 
+    def is_feasible(self, x):
```
It doesn't sound familiar to me. Feasible point, feasible region, but feasible constraint --- I'm not sure if it is a correct term.

How about `is_satisfied`?

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
         self.keep_feasible = keep_feasible
 
+    def is_feasible(self, x):
+        """
```
The formatting is sort of unconventional. I'm not sure if it very important, but I would stick to PEP8 convention starting with one-line summary.

Something like, "Check if the constraint is satisfied.". I guess that's enough.

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
+
+        Returns
+        -------
+        feasible : bool
```
If removing "feasible", then "satisfied".

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
+
+    def excess_violation(self, x):
+        """
+        How much the constraint is exceeded by.
```
Same note about conventions.

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
+
+        return feasible_lb and feasible_ub
+
+    def excess_violation(self, x):
```
If you don't like "excess_violation", I think just "violation" is not worse than that.

antonior92 replied:
I agree with Nikolay here...

andyfaff replied:
These comments aren't explicit, but it looks like you'd both prefer "violation".

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
+        Returns
+        -------
+        feasible : bool
+            Whether the vector of independent variables satisfies the 'm'
```
Mentioning 'm' here looks a bit odd. Maybe just "... satisfies the constraint." or "... all the constraints."

nmayorov commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,57 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
+            sup.filter(UserWarning)
+            ev = self.fun.fun(np.asarray(x))
+
+        excess_lb = self.bounds[0] - ev
```
Not really important, but `np.maximum` looks useful here `excess_lb = np.maximum(self.bounds[0] - ev, 0)`.

andyfaff replied:
Good catch

andyfaff commented:
CI failure is a timeout.

antonior92 commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -233,6 +244,19 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
+    >>>
+    >>> # the sum of x[0] and x[1] must be less than 1.9
+    >>> nlc = NonlinearConstraint(constr_f, -np.inf, 1.9)
+    >>> bounds = [(0,2), (0, 2)]
```
Could you use ``Bounds`` here?
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds

I think we should try to use the new classes in the documentation

andyfaff replied:
`differential_evolution` has a `bounds` positional argument which can either take the traditional bounds tuple list, or the new style `Bounds` (only added in #9945). The `bounds` positional argument always applies, any `constraints` are applied on top of that. However, to answer your point, yes `Bounds` can be used here.

andyfaff replied:
Modified that example to use a `Bounds` object.

tylerjereddy commented on scipy/optimize/_constraints.py:
```diff
@@ -249,6 +250,29 @@ def __init__(self, constraint, x0, sparse_jacobian=None,
         self.bounds = (lb, ub)
         self.keep_feasible = keep_feasible
 
+    def violation(self, x):
```
I think this function can be removed now, right? Since I merged the cross-linked pre-requisite PR recently

tylerjereddy commented on scipy/optimize/tests/test_constraints.py:
```diff
@@ -76,12 +76,21 @@ def test_prepare_constraint_infeasible_x0():
     bounds = Bounds(lb, ub, enforce_feasibility)
     pytest.raises(ValueError, PreparedConstraint, bounds, x0)
 
+    pc = PreparedConstraint(Bounds(lb, ub), [1, 2, 3])
```
the additions in this test file were also merged separately elsewhere

tylerjereddy commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -554,3 +555,99 @@ def test_converged(self):
         solver.solve()
         assert_(solver.converged())
+
+    def test_constraint_setup(self):
```
this unit test is pretty long -- might benefit from breaking into discrete tests, although not critical I suppose

andyfaff replied:
All the tests in there are looking at various numeric combinations for examining `_constraint_violation_fn` or `_calculate_population_feasibilities`; they're logically grouped, it's just that there are several of them.

tylerjereddy replied:
I understand that they're logically grouped, but at some point grouping of several smaller functions within a test class is warranted to improve reporting of failures

andyfaff commented:
@tylerjereddy, there were a couple of method calls that needed to be renamed after #9990 got merged. I've marked conversations that have been addressed as resolved (which is all of them).

tylerjereddy commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -163,6 +164,14 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
         .. versionadded:: 1.2.0
 
+    constraints : {NonLinearConstraint, LinearConstraint, Bounds}
+        Constraints on the solver, over and about those applied by the `bounds`
```
shouldn't "over and about" be "over and above" ?

andyfaff replied:
Yes

tylerjereddy commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -544,6 +578,42 @@ def __init__(self, func, bounds, args=(),
+                    # guaranteed that any x0 is feasible
+                    c.keep_feasible = False
+                    self.constraints.append(PreparedConstraint(c, x0))
+                except Exception as e:
```
the exception block is not covered by tests

andyfaff replied:
Exception would be raised if the PreparedConstraint construction doesn't work. I'd expect that to be covered by those tests.

andyfaff replied:
This is tested [here](https://github.com/scipy/scipy/blob/v1.2.1/scipy/optimize/tests/test_constraints.py#L77). I catch all `Exception`s here because I'm modifying the users constraint object, and want to definitely restore it. However, I'd only expect a `ValueError` to be raised.

tylerjereddy commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -544,6 +578,42 @@ def __init__(self, func, bounds, args=(),
+            constraints.keep_feasible = False
+            try:
+                self.constraints = [PreparedConstraint(constraints, x0)]
+            except Exception as e:
```
likewise for this exception block

andyfaff replied:
See previous comment.

tylerjereddy commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -757,11 +833,21 @@ def solve(self):
+            DE_result.maxcv = DE_result.constr_violation
+            if DE_result.maxcv > 0:
+                # if the result is infeasible then success must be False
+                DE_result.success = False
```
`DE_result.success = False` never happens in tests

andyfaff replied:
A test could be added where it's going to be impossible for constraints to be satisfied.

andyfaff replied:
Added a test for this.

tylerjereddy commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -800,12 +886,74 @@ def _calculate_population_energies(self, population):
+        else:
+            # no solution was feasible, use 'best' infeasible solution, which
+            # will violate constraints the least
+            l = np.argmin(np.sum(self.constraint_violation, axis=1))
```
This is never hit in tests -- the `if` is always True

andyfaff replied:
I can write a test for this, but I expect it will be very rare for this condition to be hit in practice, except if the constraints are really tough.

andyfaff replied:
addressed

mdhaber commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -823,6 +971,49 @@ def __del__(self):
+        accepted : bool
+
+        """
+        if feasible_orig and feasible_trial:
```
If ternary is acceptable could consider:
```
if feasible_trial:
    return energy_trial < energy_orig if feasible_orig else True
elif (cv_trial <= cv_orig).all():
    return True
```

andyfaff commented:
I'm not going to be able to make any changes until next Monday, perhaps the milestone needs to be bumped.

tylerjereddy commented:
I'll bump it, but let's get this in soon--my review notes show very few reasons not to merge at this point apart from not having an expert sign off on review.

mdhaber commented:
@andyfaff I could add as tests the benchmark problems from [Lampinen](https://pdfs.semanticscholar.org/088e/60df2694230e8e9e841e19ec218cddba54fe.pdf)

andyfaff commented:
@mdhaber, perhaps add a few - at least one LinearConstraint in the tests, and a different example in the docstring?

mdhaber commented:
@andyfaff submitted PR to your fork. Is that the best way to do this?

mdhaber commented:
Probably going to want to reformat most of these as benchmarks. There are at least two, though, that are fast enough to be tests.

andyfaff commented:
@mdhaber made some comments on your [PR](https://github.com/andyfaff/scipy/pull/10). For the additions to be scipy tests they shouldn't take more than about ~10s, a couple of minutes is too long. 

A lot of the test tolerances are set quite loosely, and I'm surprised that `differential_evolution` doesn't get closer. This may merit further inspection (but I couldn't access a lot of the Lampinen citations, so that will be hard). Possible explanations include:

1. the minimisation is stochastic (perhaps seeding is necessary for testing). Thus, for very hard problems a certain proportion of runs could be expected to 'fail'.
2. the total population sizes are different. In the Lampinen paper `Np` refers to the total population size. In the scipy implementation `popsize` is a multiplying factor for the total population size - the total population size is `np.size(x) * popsize`.
3. the way that the population is initialised could be important. scipy uses a Latin Hypercube approach to sample the available space uniformly.
4. the `updating` kwd could play an important role here. The scipy default is to continually update the best solution, whereas once per generation is the traditional approach (`updating='deferred'`).
5. the default mutation strategy in scipy is 'DE/best/1/bin', whereas the strategy in the Lampinen paper is 'DE/rand/1/bin'. To use the same mutation strategy one would need to use the `strategy='rand1bin'` kwd. The 'best' in 'best1bin' means that a trial vector is constructed by addition to the best solution so far; I conjecture (on no strong basis) that this may not be a good approach for a strongly constrained system.
6. The Lampinen paper makes multiple runs:
  > For each problem, the worst, the average and the best objective function values out of 1000 independent runs were recorded.

and

> In all cases the reported optimum solution were found.
whereas the scipy tests are single run through.

andyfaff commented:
It may be just as well the PR wasn't merged for 1.3.0. I've just figured out a large slowdown in the constraints implementation, when `NonLinearConstraint` is employed. This type of constraint eventually calls `VectorFunction.fun`, which also [implicitly evaluates](https://github.com/scipy/scipy/blob/master/scipy/optimize/_differentiable_functions.py#L144) the Hessian and Jacobian of the constraint function with finite differences. This is totally un-needed (taking a lot of time), all that's needed is the value of the evaluated function. This implicit evaluation only occurs if the `hess` parameter for `NonLinearConstraint` is an instance of `HessianUpdateStrategy`.

There are a few ways around this:

1. If the constraint supplied to `differential_evolution` is a `NonLinearConstraint` call the constraint directly without sending it through `PreparedConstraint`.
2. Temporarily change the `hess` attribute of the `NonLinearConstraint` to something that won't cause implicit updating of the Hessian and Jacobian when a `PreparedConstraint` is used.

I think option 1 is cleanest. In fact it may be cleanest to write a private version `PreparedConstraint`/`VectorFunction`/`LinearVectorFunction`/`IdentityVectorFunction` that simply allow the constraint to be evaluated without further calculation.

andyfaff commented:
Use of `_ConstraintWrapper` class means that `PreparedConstraint` can be avoided. Speedups from:

1. Not having implicit calculation of constraint Jacobian and Hessian
2. Not having to suppress warnings.

mdhaber commented:
I was wondering whether that was happening because occasionally I'd get notifications that a LinearConstraint would be more appropriate if the Jacobian is constant. Nice find. That should really speed things up!

mdhaber commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -823,6 +957,49 @@ def __del__(self):
+
+        """
+        if feasible_orig and feasible_trial:
+            return energy_trial < energy_orig
```
Technically in Lampinen (Equation 9 and in English below) this is `<=`

andyfaff replied:
Yes. `<` has been employed in `scipy` up until now. However, going back to many of the papers that I've got on DE (including Lampinen) they all use `<=`. This seems to be an error in the scipy implementation since it was first included. However, a pragmatic assessment of it's effect would indicate that it's unlikely to have a big effect.

andyfaff replied:
This will be a backwards incompatible change, but it's technically fixing a bug. Anyone relying on tests producing specific numeric output is going to be disappointed.

mdhaber commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -823,6 +957,49 @@ def __del__(self):
+            return energy_trial < energy_orig
+        elif feasible_trial and not feasible_orig:
+            return True
+        elif not feasible_trial and (cv_trial <= cv_orig).all():
```
Even though constraints can be `<=` or `>=`, this is expressed the same as in Equation 9. This is valid because of the way the constraint violations `cv` are defined. (If the constraint is violated, `cv` = the magnitude of the constraint violation, otherwise it is zero.) Looks right.

Note that Lampinen is not 100% self-consistent here. At the bottom of page 3 "Then the candidate vector can be considered less infeasible, and thus better than the current vector, if it it does not violate any of the constraints more than the current vector _and if it violate[s] less a least one of the constraints_." That last bit is not reflected in Equation 9 or the English translation below.

andyfaff replied:
Not sure what to do if the paper isn't consistent. If one considers that a trial is accepted if `energy_trial <= energy_orig`, then perhaps *"and if it violate[s] less a least one of the constraints"* can be disregarded for the time being?

mdhaber replied:
I think so. I think the simple <= here fits with the philosophy of <= for the objective function anyway.

mdhaber commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -861,18 +1045,37 @@ def __next__(self):
+
+                    # if the trial candidate is also better than the best
+                    # solution then promote it.
+                    if self._accept_trial(energy, feasible, cv,
```
Based on Lampinen alone, I'm not sure what promoting is all about. Lampinen specifically writes: "The interesting point concerning DE's replacement scheme is that a trial vector is only compared to one individual, not to all the individuals in the current population." Maybe this is why you suggested using the "deferred" option? I'll have to take a more careful look at how promotion works.

andyfaff replied:
Refer to equation 5. In the rand1bin strategy `r1, r2, r3` are randomly selected. In the best1bin strategy `r3` is not randomly selected, it's the best solution so far. 

In this implementation the best solution is kept in the first population entry. The `_promote_lowest_energy` method finds the best solution so far and moves it into the first population entry, with the previous best solution moving into its place.

If `updating=immediate` then if the trial vector is better than the best solution so far then it's immediately promoted to the best solution so subsequent trial vectors within the same generation can take advantage of the best solution.
If `updating=deferred` then the best solution is only updated once per generation.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -163,6 +165,14 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
 
         .. versionadded:: 1.2.0
 
+    constraints : {NonLinearConstraint, LinearConstraint, Bounds}
```
I originally used LBFGSB for the polishing function as it was the only (non stochastic) minimiser available at the time that could use 'simple' bounds. It would've been possible to use SLSQP, but I would've had to have built constraint dictionaries, which was a step too far for me at that time. I agree, it should be possible to use `trust-constr` but I'd prefer to leave that to a different PR.

andyfaff replied:
Ok, can use `trust-constr` to polish now. I'm not going to add SLSQP at the moment because I don't think it can use new-style constraints.
I'm not overly happy with `trust-constr` anyway, on one of the tests where there is no feasible solution `trust-constr` allows the solution to go out of bounds and still reports the polishing is a success. Anytime the solution is out of bounds, or doesn't satisfy constraints then it's essential that`OptimizeResult.success` be `False`.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -172,7 +182,9 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
-        OptimizeResult also contains the ``jac`` attribute.
+        OptimizeResult also contains the ``jac`` attribute.  If the eventual
+        solution does not satisfy the applied constraints ``success`` will be
+        `False`.
```
No polishing is done if there are constraints. Addition of that is best left to a future PR.

andyfaff replied:
> if DE does not find a solution that satisfies the constraints, attempt to polish from the least infeasible solution found, but raise a warning to alert the user of what's going on.

Done

> if DE does find a solution that satisfies the constraints but polishing is unsuccessful, return the DE optimize result with an appropriate status message (DE successful, polishing unsuccessful).

I think it's just better to ignore the polishing result in this case.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -233,6 +245,20 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
+    Let's try and do a constrained minimization
+    >>> from scipy.optimize import NonlinearConstraint, Bounds
+    >>> def constr_f(x):
+    ...     return np.array(x[0] + x[1])
```
> I suppose once you fix the jac/hess evaluation issue there will be no inherent speed advantage to using LinearConstraint?

I think this is the case, yes.

> In this case, for example, using a NonlinearConstraint with return np.sum(x) would probably be even faster than a LinearConstraint with A = [1, 1], right?

I would assume so.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -800,12 +872,74 @@ def _calculate_population_energies(self, population):
+        """
+        return np.concatenate([c.violation(x) for c in self.constraints])
+
+    def _calculate_population_feasibilities(self, population):
```
The population consists of values between 0 and 1. Scale/unscaling of parameters uses the lower and upper bounds to transform those normalised values to an actual parameter value.

mdhaber replied:
I figured but was trying to check that code, too.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -1026,3 +1244,86 @@ def __init__(self, f, args):
+        elif isinstance(constraint, Bounds):
+            def fun(x):
+                A = np.eye(len(x))
+                return A.dot(x)
```
This was just copied from `PreparedConstraint`/`IdentityVectorFunction`. It could just return `x`.

mdhaber replied:
@antonior92 any reason to multiply by identity here?

andyfaff replied:
I'll just return `x` here.

andyfaff commented:
@mdhaber , would you mind if a squashed the commits in here? 1 commit from before you added the tests, 1 commit for your test addition, 1 commit for changes made after your addition.

mdhaber commented:
Ok.

andyfaff commented:
I've tidied up the commit history, this PR needed squashing. I think it's in a position to merge.

mdhaber commented:
Restarted test with build failure. 
I should check this out one more time before merging. So are all the tests fast enough now? Getting the accuracy you expected?

andyfaff commented:
Sorry, the individual commits were lost when I squashed in preparation for a merge.

Test changes
--------------
1. `x_opt` was added to all of the tests to check `f_opt` vs `f(x_opt)`.
2. All the tests use the `rand1bin` strategy.
3. All the tests are seeded for reproducibility.
4. One test was marked as slow
5. the default for `maxiter` was used.
6. smaller `popsize`s (corresponding to the Lampinen paper) were used. The minimisation is proportionally speeded up. All the tests complete in under 3 secs, apart from L8.
7. Because the differential evolution doesn't get close enough to the optimal solution without a large number of iterations (due to the convergence statistic being satisfied too early - discussed in an earlier conversation), I introduced the use of `trust-constr` minimisation as a polishing routine. This gets the solution quite close to the optimal. More iterations can be requested by setting the tolerances more tightly.
8. the polishing enabled the test tolerances to be set much tighter, all tests pass but can be relaxed if there are future changes.

differential_evolution code
---------------------------

1. A polish using `trust-constr` is now done.
2. `PreparedConstraint` was discarded for a plain `_ConstraintWrapper`, which allows simple evaluation of constraints without the Jacobian of the constraints being calculated. This is a big speedup.
3. the accept test inequality was fixed (when both solutions were feasible)

mdhaber commented:
I haven't forgotten! I'll try to get to this soon.

mdhaber commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -554,3 +559,576 @@ def test_converged(self):
+        res = differential_evolution(f, bounds, strategy='rand1bin', seed=1234,
+                                     constraints=constraints)
+        x_opt = (14.095, 0.84296)
+        f_opt = -6961.814744
```
I was playing around with this one because the tolerance was not very tight. `slsqp` found the optimum reported by Lampinen. Funny, though - I tried manually polishing with `trust-constr` and got a different value than the auto-polishing by DE. Thoughts on why that might happen?

andyfaff replied:
If the starting points were the same (unpolished DE), then I don't know why.

andyfaff replied:
I can get to the best minimum if I set the de tolerance to be very low. As you said, the polishing done in DE isn't getting there. The DE polish is only accepted if the polish is successful, the solution lies strictly within the bounds, and polish.fun is less than de.fun. Perhaps that's why the polishing isn't getting there.

mdhaber commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -554,3 +559,576 @@ def test_converged(self):
+                                     constraints=constraints)
+
+        x_opt = (78.0, 33.0, 29.995, 45.0, 36.776)
+        f_opt = -30665.608768
```
Strange - the value reported in Lampinen is much lower than this one, which was taken from Koziel. But Koziel reports this value here `-30665.608768` as "the optimum", whereas he is careful to note others as "the best known value". I wonder if Lampinen made a mistake when implementing this one.

mdhaber replied:
DE is getting ~ -30665.608768 so there's nothing to be changed about the test. Just weird that Lampinen reports a different value.

andyfaff replied:
For this particular example Koziel actually reports -30665.5. However, f(x_opt) is -30665.608768 - his evaluated function isn't correct.

Having said that I can't get to the actual minimum, the closest I can get is -30665.538, when `res.x` is within `atol=1e-3` of `x_opt`.

Investigating this further I get `[92.000043492964, 98.8405108381325, 19.999935068412004]` when evaluating the NonlinearConstraint `N.fun(x_opt)`. The first and last values are outside the constraint bounds `[(0, 92), (90, 110), (20, 25)]`. Thus, the reported `x_opt` in Koziel doesn't satisfy the constraints. I wonder if he was using single precision or something.

I'm going to fix this by using our values, coz I trust ours more.

mdhaber commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -554,3 +559,576 @@ def test_converged(self):
+                                     seed=1234, constraints=constraints,
+                                     popsize=3)
+
+        f_opt = 7049.330923
```
And DE without polishing is getting very close to this answer when it is forced to iterate more. That's great.

andyfaff replied:
Koziel is higher, Lampinen finds a lower minimum. Changed the code to the better minimum

andyfaff replied:
Tolerances now much tighter.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -163,6 +166,14 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
+        Constraints on the solver, over and above those applied by the `bounds`
+        kwd. Uses the approach by Lampinen [5]_. There is no particular
+        advantage employing `LinearConstraint` over `NonLinearConstraint` as no
+        Jacobian/Hessian of the constraint functions is used.
```
There's no advantage for the DE, but the polishing can benefit from it.

mdhaber replied:
Hmm, true.

andyfaff replied:
I decided to get rid of that bit. Whilst true for the de part, the polishing would still find it useful. I think in the long run it'll confuse too many users.

andyfaff commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -554,3 +559,576 @@ def test_converged(self):
+            np.array([[0.5, 0.5], [1., 1.]]))
+        assert_equal(feas, [False, False])
+        assert_almost_equal(cv, np.array([[0.1], [2.1]]))
+        assert cv.shape == (2, 1)
```
AFAIR `assert_` needed to be used when we used `nose`. Now we use `pytest` one can use plain `assert`.

andyfaff commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -554,3 +559,576 @@ def test_converged(self):
+        def f(x):
+            x = np.hstack(([0], x))  # 1-indexed to match reference
+            fun = (np.sin(2*np.pi*x[1])**3*np.sin(2*np.pi*x[2]) /
+                   (x[1]**3*(x[1]*x[2])))
```
I fixed the equation but can't get lower than -0.0958

mdhaber replied:
That's right! Lampinen reports `.095825` as the solution to the maximimization problem, so what you are getting is the correct solution to the minimization. The desired value in the test should be changed.

andyfaff replied:
Koziel is higher. This is what I meant when I said that often the benchmarks in papers aren't all checked/correct.

andyfaff commented:
If this passes now I think it's good to go.

andyfaff commented:
We're all good now. The pypy3 test fail can be ignored.

andyfaff commented:
@mdhaber, could we merge this?

andyfaff commented:
@mdhaber thanks for your sterling review and test development work on this.

mdhaber commented:
NP and thanks for the PR. In the past, I've only used local minimizers for nonlinear problems (mostly optimal control), so it was great to play around with DE for the first time. I was impressed!
